store/store.go

```diff
@@ -245,82 +275,181 @@ func (bs *BlockStore) LoadSeenCommit(height int64) *types.Commit {
+		panic(err)
+	}
+	defer iter.Close()
+
 	batch := bs.db.NewBatch()
 	defer batch.Close()
-	flush := func(batch dbm.Batch, base int64) error {
-		// We can't trust batches to be atomic, so update base first to make sure noone
-		// tries to access missing blocks.
-		bs.mtx.Lock()
-		bs.base = base
-		bs.mtx.Unlock()
-		bs.saveState()
-
-		err := batch.WriteSync()
-		if err != nil {
-			return fmt.Errorf("failed to prune up to height %v: %w", base, err)
-		}
-		batch.Close()
-		return nil
-	}
 
-	for h := base; h < height; h++ {
-		meta := bs.LoadBlockMeta(h)
-		if meta == nil { // assume already deleted
-			continue
-		}
-		if err := batch.Delete(blockMetaKey(h)); err != nil {
-			return 0, err
+	pruned := uint64(0)
+	for iter.Valid() {
+		// delete the block meta
+		if err := batch.Delete(iter.Key()); err != nil {
+			return 0, fmt.Errorf("pruning error at key %X: %w", iter.Key(), err)
 		}
-		if err := batch.Delete(blockHashKey(meta.BlockID.Hash)); err != nil {
-			return 0, err
+
+		// load the block meta so we can also remove the block hash from the db
+		var pbbm = new(tmproto.BlockMeta)
+		bz := iter.Value()
+
+		err = proto.Unmarshal(bz, pbbm)
+		if err != nil {
+			panic(fmt.Errorf("unmarshal to tmproto.BlockMeta: %w", err))
 		}
-		if err := batch.Delete(blockCommitKey(h)); err != nil {
-			return 0, err
+
+		blockMeta, err := types.BlockMetaFromProto(pbbm)
+		if err != nil {
+			panic(fmt.Errorf("error from proto blockMeta: %w", err))
 		}
-		if err := batch.Delete(seenCommitKey(h)); err != nil {
-			return 0, err
+
+		if err := batch.Delete(blockHashKey(blockMeta.BlockID.Hash)); err != nil {
+			return 0, fmt.Errorf("pruning error at key %X: %w", iter.Key(), err)
 		}
-		for p := 0; p < int(meta.BlockID.PartSetHeader.Total); p++ {
-			if err := batch.Delete(blockPartKey(h, p)); err != nil {
-				return 0, err
+
+		pruned++
+		// avoid batches growing too large by flushing to database regularly
+		if pruned%1000 == 0 {
+			if err := iter.Error(); err != nil {
+				return pruned, err
+			}
+			if err := iter.Close(); err != nil {
+				return pruned, err
+			}
+
+			err := batch.Write()
+			if err != nil {
+				return pruned, fmt.Errorf("pruning error at key %X: %w", iter.Key(), err)
+			}
+			if err := batch.Close(); err != nil {
+				return pruned, err
 			}
+
+			iter, err = bs.db.Iterator(
+				blockMetaKey(1),
+				blockMetaKey(retainHeight),
+			)
+			if err != nil {
+				panic(err)
+			}
+			defer iter.Close()
+
+			batch = bs.db.NewBatch()
+			defer batch.Close()
+		} else {
+			iter.Next()
+		}
+	}
+	if err := iter.Error(); err != nil {
+		return pruned, err
+	}
+
+	err = batch.WriteSync()
+	if err != nil {
+		return pruned, fmt.Errorf("pruning error at key %X: %w", iter.Key(), err)
+	}
+
+	return pruned, nil
+}
+
+// batchDelete is a generic function for deleting a range of values based on the lowest
+// height up to but excluding retainHeight
+func (bs *BlockStore) batchDelete(start []byte, end []byte) error {
```

Comment:
This is quite a generic function (i.e. state store does the exact same thing). It would probably make more sense to extract this. My first thought is to add it to the tm-db? or to the lib directory. Or should I just repeat myself (I remember someone telling me there's nothing wrong with that). What are people's thoughts on this?

Reply:
Hmm actually it might be not so generic as I thought. State store has a few nuances. I will just repeat myself for now

Reply:
Might make sense to extract this out to a helper function, but maybe design it in such a way that it can handle those "special" cases too, e.g. by using callback functions.

Reply:
I guess the main reason why I would have two distinct functions is because the block store iterates forward and the state store iterates backwards (because state store has this entire dependency dynamic)